crates/noirc_evaluator/src/ssa_refactor/acir_gen/acir_ir/acir_variable.rs

```diff
@@ -331,6 +336,218 @@ impl AcirContext {
+    /// is correct according to the function definition.
+    ///
+    /// Some functions expect a variable number of inputs and in such a case,
+    /// this method will do nothing.  An example of this is sha256.
+    /// In that case, this function will not check anything.
+    ///
+    /// Since we expect black box functions to be called behind a Noir shim function,
+    /// we trigger a compiler error if the inputs do not match.
+    ///
+    /// An example of Noir shim function is the following:
+    /// ``
+    /// #[foreign(sha256)]
+    /// fn sha256<N>(_input : [u8; N]) -> [u8; 32] {}
+    /// ``
+    fn intrinsics_check_inputs(&mut self, name: BlackBoxFunc, inputs: &[AcirVar]) {
+        let expected_num_inputs = match Self::black_box_func_expected_input_size(name) {
+            Some(expected_num_inputs) => expected_num_inputs,
+            None => return,
+        };
+        let got_num_inputs = inputs.len();
+
+        assert_eq!(expected_num_inputs,inputs.len(),"Tried to call black box function {name} with {got_num_inputs} inputs, but this function's definition requires {expected_num_inputs} inputs");
+    }
+
+    /// Intrinsic calls expect their inputs to be in a specific data structure (FunctionInput).
+    ///
+    /// This function will convert `AcirVar` into `FunctionInput` for an intrinsic call.
+    fn prepare_inputs_for_intrinsics_call(
+        &mut self,
+        name: BlackBoxFunc,
+        inputs: &[AcirVar],
+    ) -> Result<Vec<FunctionInput>, AcirGenError> {
+        self.intrinsics_check_inputs(name, inputs);
+
+        let mut witnesses = Vec::new();
+        for input in inputs {
+            let var_data = &self.data[input];
+
+            // Intrinsics only accept Witnesses. This is not a limitation of the
+            // intrinsics, its just how we have defined things. Ideally, we allow
+            // constants too.
+            let expr = var_data.to_expression();
+            let witness = self.acir_ir.get_or_create_witness(&expr);
+
+            // Fetch the number of bits for this variable
+            // If it has never been constrained before, then we will
+            // encounter None, and so we take the max number of bits for a
+            // field element.
+            let num_bits = match self.variables_to_bit_sizes.get(input) {
+                Some(bits) => {
+                    // TODO: First lets check to see how these are handled by the IR
+                    //
+                    // In Noir, we specify the number of bits to take from the input
+                    // by doing the following:
+                    //
+                    // ```
+                    // call_intrinsic(x as u8)
+                    // ```
+                    //
+                    // The `as u8` specifies that we want to take 8 bits from the `x`
+                    // variable.
+                    //
+                    // TODO(Jake): There were discussions about the SSA IR optimizing out range
```

Comment:
@jfecher pinging for your insight here

Reply:
We could consider removing the actual cast instruction and reinterpreting the type of x without inserting an extra, unnecessary range constraint. So I think there are ways to keep this optimization and work around this problem, but it is also not the most important optimization so it is just something to keep in mind for the future.

Reply:
Created issue #1467 to track this